source/solvers/fluid_dynamics_matrix_free.cc

```diff
@@ -2544,27 +2555,43 @@ FluidDynamicsMatrixFree<dim>::calculate_time_derivative_previous_solutions()
     }
 }
 
+template <int dim>
+void
+FluidDynamicsMatrixFree<dim>::create_GMG()
+{
+  gmg_preconditioner = std::make_shared<MFNavierStokesPreconditionGMG<dim>>(
+    this->simulation_parameters,
+    this->dof_handler,
+    this->dof_handler_fe_q_iso_q1);
+
+  gmg_preconditioner->reinit(this->mapping,
+                             this->cell_quadrature,
+                             this->forcing_function,
+                             this->simulation_control,
+                             this->fe);
+}
+
+template <int dim>
+void
+FluidDynamicsMatrixFree<dim>::initialize_GMG()
+{
+  dynamic_cast<MFNavierStokesPreconditionGMG<dim> *>(gmg_preconditioner.get())
+    ->initialize(this->simulation_control,
+                 this->flow_control,
+                 this->present_solution,
+                 this->time_derivative_previous_solutions);
+}
```

Comment:
These are (new) function that can be overridden in the derivated classes (e.g., to create a different type of multgrid type and to pass additional information to the initialization methods).